ros2/ament_setup.cpp.tpl

```diff
@@ -0,0 +1,39 @@
+#include "{{header}}"
+
+#include <cstdlib>
+#include <string>
+
+namespace {{namespace}} {
+namespace {
+
+constexpr const char* AMENT_PREFIX_PATH_ENV = "AMENT_PREFIX_PATH";
+constexpr const char* AMENT_PREFIX_PATH = "{{ament_prefix_path}}";
+
+struct AmentPrefixPathSetup
+{
+  AmentPrefixPathSetup()
+  {
+    const char* ament_prefix_path = std::getenv(AMENT_PREFIX_PATH_ENV);
+    if (ament_prefix_path)
+    {
+      std::string ament_prefix_path_str = ament_prefix_path;
+      ament_prefix_path_str += ":";
+      ament_prefix_path_str += AMENT_PREFIX_PATH;
+      setenv(AMENT_PREFIX_PATH_ENV, ament_prefix_path_str.c_str(), 1);
+    }
+    else
+    {
+      setenv(AMENT_PREFIX_PATH_ENV, AMENT_PREFIX_PATH, 1);
+    }
```

Comment:
This is different compared to the logic in sh launcher. What's the reasoning here? Or a use-case?

Reply:
As-is, this code allows other prefix paths to sneak in -- like from a regular ROS installation. That opens a can of worms and is a no-go for me at the moment.

I would expect one to use a single ros2_ament_setup_library per one binary. All deps given to ros2_ament_setup are run through machinery that will collect necessary info from all given deps and form an ament-compatible file tree that's going to be used by the app.

An alternative is to refactor your ros2_ament_setup_library to have have

```
        "ament_setups": attr.label_list(
            mandatory = True,
            providers = [Ros2AmentSetupInfo],
        ),
```
Then you can merge `DefaultInfo`s and form a unique prefix path e.g. `"path1:path2:...:pathn"` that's going to be baked in the .cc file. Keep in mind that in this case is expected there will be potentially massive amount of symlinks from different prefix path folders to same files. That's why I would go for option 1 and use one ros2_ament_setup_library per binary.

If you really need option 2 right now, please tell me more about your use-case.

Please let me know what you think.

Reply:
> As-is, this code allows other prefix paths to sneak in -- like from a regular ROS installation. That opens a can of worms and is a no-go for me at the moment. I would expect one to use a single ros2_ament_setup_library per one binary. 

Could you explain the can of worms involved in doing this? Obviously sourcing regular ROS prior to running bazel-based applications is always a bad idea (I already found that  `@ros2cli` doesn't work correctly after sourcing ROS). The idea of this logic is that it's actually fine to combine the `AMENT_PREFIX_PATH` from multiple sources (e.g. one library that depends on rosbag plugins, another library that depends on some specific message types and so on) and we just keep expanding the `AMENT_PREFIX_PATH` as the relevant C++ code needs access to those types. Yes, you might end up with multiple instances of the same plugin libraries in the runfiles (each in different locations), but since the compilation options and source code is supposed to be the same within one bazel invocation, they should be equivalent and it doesn't matter which one is used.

The value I see in being able to do this, is that you can restrict the "ros-awareness" to specific cc_libraries. Each library just does what it needs to work correctly (this is a form of encapsulation). If you wanted to restrict the number of `ros2_ament_setup_library`s per binary, you will have to somehow use ros2-specific logic at the top level (i.e. the rule creating the executable) to confirm that. Or worse, not check it and silently run into hard-to-debug issues when you somehow do end up with multiple. In most cases such top-level logic is fine (ros2_cpp_binary actually works great!), but in my case I would really like to keep the ros-specific part encapsulated.

I can understand that this is not a super common use case, but other than the possible duplication of some symlinks in the runfiles I don't see a big problem with it.

Reply:
> You just said there is a potential use-case where something might fail. :)

Is running with pre-sourced ROS something you want to support?

> > I can understand that this is not a super common use case, but other than the possible duplication of some symlinks in the runfiles I don't see a big problem with it.
> 
> OK, would my option 2 work for you?

If I understand your option 2 correctly, the problem I see with it is that you still need a single location where you bake the single merged path into the binary, which breaks the encapsulation.

Reply:
> Is running with pre-sourced ROS something you want to support?

This opens doors to various kind of problems, you already mentioned one. So, unless there is a general consensus among the users of the repo that we should do that, my answer is no.

Reply:
I totally agree that mixing stock ROS and `rules_ros2` is not a good idea.

@woutervdstoel The issue you saw with `ros2cli` was potentially due to the `PYTHONPATH` introduced by sourcing the stock ROS installation. Explicitly unsetting `PYTHONPATH` before running Bazel/`rules_ros2` targets fixed this for me.

As for the can of worms that appending to `AMENT_PREFIX_PATH` opens, I'm not sure. However, I would also argue for the approach where the final value for `AMENT_PREFIX_PATH` is assembled in one place, as that is more in line with the Bazel philosophy, where you give the build system as much information as possible and don't rely so much on assembling stuff at runtime. I agree that a single `ros2_ament_setup` per binary should be sufficient. By having a single ROS-aware top level rule that pulls all deps and takes care of ament setup you still have encapsulation. But you don't hide the fact that this somehow depends on ROS, which is actually a good thing IMHO. Too much magic also often can lead to hard-to-debug issues. The `ros2_ament_setup` also doesn't need to happen on the level of the binary. You can have a `cc_library` take care of that. If you really want to hide the fact that ROS is involved, you could also use custom versions of `cc_library` and `cc_{binary,test}` that transparently pass along the `Ros2AmentSetupInfo`s and only on the binary level create the actual `ros2_ament_setup` target. I'm not sure how much effort that would be, though.

The underlying issue here is that the natural way for a Bazel-based setup is very different from the native ROS way. Stock ROS relies heavily on resolving dependencies at runtime, while in a Bazel setup you want to decide as much as possible at build time. One way around this is creating package rules that produce packages in the native ROS format, including setup scripts. Over at [ApexAI/rules_ros](https://github.com/ApexAI/rules_ros/) that seems to be the idea. But with this approach you would lose some of the Bazel advantages such as being able to `bazel run` fine-grained targets. Instead, you would need one `ros2` target that depends on everything if you wanted to mimic the stock `ros2` behavior.

Overall, I think staying with a single `ros2_ament_setup` per binary would be more in line with the philosophy of `rules_ros2`. And I would trust my users can handle being exposed to the fact that ROS is used underneath. If you decide to go with the approach that involves more magic, make sure to cover that very well with tests. Otherwise that next ROS update might break something and leave you with that hard-to-debug issue. If you want encapsulation to be able to easily change ROS for something else some time in the future, I would say, forget it; that'd be a major refactor either way. ;-)

One last thought how one could work around the issue of some external `AMENT_PREFIX_PATH` sneaking in: You could maintain a "private" version (via another env variable or a static variable inside the executable) and use that to set `AMENT_PREFIX_PATH`, disregarding any value that `AMENT_PREFIX_PATH` had previously. However, since that would involve quite some unexpected "magic", I would most certainly not go for it...

Reply:
I fully agree with your statement that the native ROS2 plugin system does not play nicely with the bazel principle of specifying dependencies at build time. I also agree that trying to hide this fact using the approach from this PR is always going to result in a leaky abstraction. But I don't think that should be a reason not to allow it.

Forcing the usage of specific rules all the way to the top level of the dependency graph only to use a ROS library in some implementation detail is also a drawback, and I think that depending on the usage one negative can outweigh the other. I fully agree that in most cases it's better to use a specific top-level rule (or a single `ros2_ament_setup` per binary) but I still believe there's a use case for appending if one understands the risks.

This is why I just pushed a proposed change to make the default behavior to override the `AMENT_PREFIX_PATH` and the append-behavior opt-in. I also covered this with some unit tests.

It turns out the plugin loader only accepts one plugin providing types in a specific namespace so I also made the namespaces unique.